examples/async_fl/async_aggregator.py

```diff
@@ -95,54 +105,29 @@ def aggregate_client_weights(self, results):
             param_weight = results['update_weight'][p]
+
             if isinstance(param_weight, list):
                 param_weight = np.asarray(param_weight, dtype=np.float32)
             param_weight = torch.from_numpy(
                 param_weight).to(device=self.device)
 
-            if self.model_in_update == 1:
-                self.model_weights[p].data = param_weight * importance
-            else:
+            if self.model_weights[p].data.dtype in (
+                torch.float, torch.double, torch.half, 	
+                torch.bfloat16, torch.chalf, torch.cfloat, torch.cdouble
+            ):
+                # Only assign importance to floats (trainable variables) 
                 self.model_weights[p].data += param_weight * importance
+            else:
+                # Non-floats (e.g., batches), no need to aggregate but need to track
+                self.model_weights[p].data += param_weight
 
         if self.model_in_update == self.async_buffer_size:
+            logging.info("Calibrating tensor type")
             for p in self.model_weights:
-                d_type = self.model_weights[p].data.dtype
-
-                self.model_weights[p].data = (
-                    self.model_weights[p] / float(self.async_buffer_size)).to(dtype=d_type)
-
-    def aggregate_client_group_weights(self, results):
-        """Streaming weight aggregation. Similar to aggregate_client_weights,
-        but each key corresponds to a group of weights (e.g., for Tensorflow)"""
-
-        client_staleness = self.round - \
-            self.client_model_version[results['clientId']]
-        importance = 1. / math.sqrt(1 + client_staleness)
-
-        for p_g in results['update_weight']:
-            param_weights = results['update_weight'][p_g]
-            for idx, param_weight in enumerate(param_weights):
-                if isinstance(param_weight, list):
-                    param_weight = np.asarray(param_weight, dtype=np.float32)
-                param_weight = torch.from_numpy(
-                    param_weight).to(device=self.device)
-
-                if self.model_in_update == 1:
-                    self.model_weights[p_g][idx].data = param_weight * importance
-                else:
-                    self.model_weights[p_g][idx].data += param_weight * importance
-
-        if self.model_in_update == self.async_buffer_size:
-            for p in self.model_weights:
-                for idx in range(len(self.model_weights[p])):
-                    d_type = self.model_weights[p][idx].data.dtype
-
-                    self.model_weights[p][idx].data = (
-                        self.model_weights[p][idx].data /
-                        float(self.async_buffer_size)
-                    ).to(dtype=d_type)
+                d_type = self.weight_tensor_type[p]
+                self.model_weights[p].data = (self.model_weights[p].data/float(self.async_buffer_size)).to(dtype=d_type)
```

Comment:
Dividing the model weights by `async_buffer_size` might not be correct since we are weighing each update by the staleness factor. Should we be dividing by the sum of the importance variable over the round?

Reply:
Or is `self.model_weights[p].data` the delta?

Reply:
Yes! This is really a good catch! I was wondering this too, so turned to the fedbuff paper. It turns out [the paper ](https://arxiv.org/abs/2106.06639) divides the update by async_buffer_size again (Alg 1, L11). But we are happy to update it if you think it makes more sense.

Reply:
[self.model_weights[p].data += delta * importance](https://github.com/SymbioticLab/FedScale/pull/162/files/19939ac6b5b55f394358946cd1346cc1130b5d60#diff-ceaf64e0eb8e76f1f34f2d04c7ba134994157ccca358e9b3b56d9920d76596c4L107). So at this time, it becomes the latest model version.

Reply:
I have a quick question, and would greatly appreciate your input: existing FL runs the local training over a fixed number of steps, while the step here can be iteration or epoch. We find it is difficult to have a consistent implementation, though this change only takes 2-3 lines of code (e.g., by changing local_step = len(data) * local_step).

Wondering which one you prefer. We think local_step=epoch can lead to longer round duration due to the data heterogeneity. Thanks!

Reply:
I see, I think dividing by `async_buffer_size` is fine. However, if I understand correctly, `self.model_weights[p].data = (self.model_weights[p].data/float(self.async_buffer_size)).to(dtype=d_type)` is dividing the entire model weights by `async_buffer_size`, which might not be correct? We would want to divide just the accumulated gradients of this round by the async_buffer_size, and then multiply by a learning rate after then add that to the global model weights. Again, it could be that I'm not understanding the code correctly. 😄

Reply:
As for fixed number of steps vs len(data), I think it might make sense to have a max number of steps if possible. Then this value can be adjusted for different datasets based on use case.

Reply:
> 

Oh!  Yes! You are correct! This might even be the cause of weird accuracy. Many thanks for helping me! Let me fix it and try!